### PR TITLE
Implement debugging of iOS Hermes application

### PR DESCRIPTION
**Pull request by JiglioNero**
#1480

### CODE REVIEW

File: src/debugger/debugSessionBase.ts [hunk 1]
```diff
@@ -54,6 +54,7 @@ export interface IAttachRequestArgs
     cwd: string /* Automatically set by VS Code to the currently opened folder */;
     port: number;
     url?: string;
+    useHermesEngine: boolean;
```
Comment (RedMickey) [inline src/debugger/debugSessionBase.ts:57]:
```suggestion
    useHermesEngine?: boolean;
```

File: src/debugger/direct/directDebugSession.ts [hunk 2]
```diff
@@ -273,4 +290,12 @@ export class DirectDebugSession extends DebugSessionBase {
             vscode.commands.executeCommand(this.stopCommand, this.session);
         }
     }
+
+    protected async initializeSettings(args: any): Promise<any> {
+        return super.initializeSettings(args).then(() => {
```
Comment (RedMickey) [inline src/debugger/direct/directDebugSession.ts:295]: It could be better to use async/await syntax here.

File: src/common/error/errorStrings.ts [hunk 3]
```diff
@@ -326,4 +326,8 @@ export const ERROR_STRINGS = {
         "ErrorWhileSelectMonitoringLogCat",
         "No active Android LogCat monitors found",
     ),
+    [InternalErrorCode.AndroidCouldNotDirectDebugWithoutHermesEngine]: localize(
+        "AndroidCouldNotDirectDebugWithoutHermesEngine",
+        "Could not direct debug Android without the Hermes engine. Please switch 'useHermesEngine' option to 'true'",
```
Comment (RedMickey) [inline src/common/error/errorStrings.ts:331]:
```suggestion
        "Could not start direct debugging of the Android application without Hermes engine enabled. Please check if Hermes engine is enabled in your project and your debugging configuration contains 'useHermesEngine' prop set to 'true'",
```

File: src/common/configurationProviderHelper.ts [hunk 4]
```diff
@@ -80,6 +81,35 @@ export class ConfigurationProviderHelper {
...
+        config: Partial<ILaunchRequestArgs>,
+        step: number,
+        totalSteps: number,
+    ): Promise<Partial<ILaunchRequestArgs>> {
+        let shouldUseHermes = await input.showQuickPick<
+            DebugConfigurationQuickPickItem,
+            IQuickPickParameters<DebugConfigurationQuickPickItem>
+        >({
+            title: localize("UseHermesEngine", "Use Hermes engine"),
+            placeholder: localize("UseHermesEnginePrompt", "Use Hermes engine for direct debug?"),
```
Comment (RedMickey) [inline src/common/configurationProviderHelper.ts:95]:
```suggestion
            placeholder: localize("UseHermesEnginePrompt", "Use Hermes engine for direct debugging?"),
```

File: src/extension/debuggingConfiguration/reactNativeDebugConfigProvider.ts [hunk 5]
```diff
@@ -276,13 +309,15 @@ export class ReactNativeDebugConfigProvider implements vscode.DebugConfiguration
...
                     state.config.platform,
+                    state.config.useHermesEngine !== false,
                     true,
                 );
             } else {
                 state.config.name = DebugScenarioNameGenerator.createScenarioName(
                     state.scenarioType,
                     state.config.type || DEBUG_TYPES.REACT_NATIVE,
                     state.config.platform,
+                    false,
```
Comment (RedMickey) [inline src/extension/debuggingConfiguration/reactNativeDebugConfigProvider.ts:320]:
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -115,17 +116,17 @@ Once app is loaded and running, [open the developer menu](https://reactnative.de
 
 The extension allows you to debug multiple devices and configurations, please read the following sections for more information for your particular use case.
 
-## Hermes (Android)
+## Hermes engine
 
 Hermes is the new JavaScript engine optimized for running React Native apps on Android. It improves app performance and decreases app size.
```
Comment (RedMickey) [inline README.md:121]:
```suggestion
Hermes is the new JavaScript engine optimized for running React Native apps on Android and iOS. It improves app performance and decreases app size.
```

File: README.md [hunk 7]
```diff
@@ -137,11 +138,17 @@ To debug while using Hermes engine, please choose one of the following debug con
 }
 ```
 
-- "React Native Direct: Attach to the React Native Hermes - Experimental
+### iOS Hermes
+
+The extension allows you to debug iOS Hermes applications. See [iOS Hermes debugging](#ios-hermes-debugging) for more details.
```
Comment (RedMickey) [inline README.md:143]:
```suggestion
The extension provides experimental support of debugging iOS Hermes applications. See [iOS Hermes debugging](#ios-hermes-debugging) for more details.
```

File: README.md [hunk 8]
```diff
@@ -137,11 +138,17 @@ To debug while using Hermes engine, please choose one of the following debug con
...
+
+### Attach to Hermes application
+
+To attach to the running Hermes application use `Attach to Hermes application - Experimental` launch configuration:
```
Comment (RedMickey) [inline README.md:147]:
```suggestion
To attach to a running Hermes application use `Attach to Hermes application - Experimental` launch configuration:
```

File: README.md [hunk 9]
```diff
@@ -176,7 +183,7 @@ Please be aware, specifying the scheme value as a part of the `runArguments` par
 
 ### iOS direct debugging
 
-The extension provides experimental support of iOS direct debugging. See more info here: [react-native-community/discussions-and-proposals#40](https://github.com/react-native-community/discussions-and-proposals/issues/40), [react-native-community/discussions-and-proposals#206](https://github.com/react-native-community/discussions-and-proposals/issues/206)
+The extension provides experimental support of iOS direct debugging. At the moment direct iOS debugging only works with applications running on the device. See more info here: [react-native-community/discussions-and-proposals#40](https://github.com/react-native-community/discussions-and-proposals/issues/40), [react-native-community/discussions-and-proposals#206](https://github.com/react-native-community/discussions-and-proposals/issues/206)
```
Comment (RedMickey) [inline README.md:186]:
```suggestion
The extension provides experimental support of iOS direct debugging. See more info here: [react-native-community/discussions-and-proposals#40](https://github.com/react-native-community/discussions-and-proposals/issues/40), [react-native-community/discussions-and-proposals#206](https://github.com/react-native-community/discussions-and-proposals/issues/206)

For now the extension supports iOS direct debugging only on real iOS devices.
```

File: README.md [hunk 10]
```diff
@@ -207,6 +215,27 @@ You can use the following debug scenarios to debug iOS apps directly:
     "port": 9221
 ```
 
+### iOS Hermes debugging
+
+You can enable Hermes engine for iOS application by edit `ios/Podfile` file and make the change illustrated below:
```
Comment (RedMickey) [inline README.md:220]:
```suggestion
You can enable Hermes engine for an iOS application by editing `ios/Podfile` file the following way:
```

File: README.md [hunk 11]
```diff
@@ -207,6 +215,27 @@ You can use the following debug scenarios to debug iOS apps directly:
...
+
+```json
```
Comment (RedMickey) [inline README.md:222]:
```suggestion
```diff
```

File: README.md [hunk 12]
```diff
@@ -207,6 +215,27 @@ You can use the following debug scenarios to debug iOS apps directly:
...
+-  use_react_native!(:path => config[:reactNativePath])
++  use_react_native!(:path => config[:reactNativePath], :hermes_enabled => true)
+```
+
+After this changes you need to execute `pod install` in `ios` folder. And now you can use `Debug iOS Hermes - Experimental` launch configuration to debug iOS Hermes application:
```
Comment (RedMickey) [inline README.md:227]:
```suggestion
After this change you need to execute `pod install` command in `ios` folder. After that you can use `Debug iOS Hermes - Experimental` launch configuration to debug an iOS Hermes application:
```

File: src/extension/debuggingConfiguration/debugScenarioNameGenerator.ts [hunk 13]
```diff
@@ -84,21 +86,19 @@ export class DebugScenarioNameGenerator {
...
         debugType: string,
+        useHermesEngine: boolean,
         platformType?: PlatformType | string,
     ) {
+        if (useHermesEngine) {
+            if (debugScenarioType === DebugScenarioType.AttachApp) {
+                debugScenarioName.postPlatformTypeDescription = "Hermes application";
+            } else {
+                debugScenarioName.postPlatformTypeDescription = "Hermes";
+            }
```
Comment (etatanova) [inline src/extension/debuggingConfiguration/debugScenarioNameGenerator.ts:97]: Maybe it's better to use the ternary operator here?
```suggestion
            debugScenarioName.postPlatformTypeDescription = (debugScenarioType === DebugScenarioType.AttachApp) ? "Hermes application" : "Hermes";
```

**Review by RedMickey — APPROVED** [review]

**Review by etatanova — APPROVED** [review]

**Pull request merged**